File: src/main/java/com/joseestudillo/spark/rdd/BroadcastVariables.java
package com.joseestudillo.spark.rdd;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import org.apache.log4j.Logger;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.broadcast.Broadcast;

import com.joseestudillo.spark.utils.SparkUtils;

/**
 * 
 * Broadcast variable examples
 * 
 * Broadcast vars allow you to have a read-only value in all the workers avoiding the need of sending them every time they are needed. the value of the property
 * {@code spark.serializer} affects to this, as the values of broadcast variables must be serialized to be sent to all the workers
 * 
 * @author Jose Estudillo
 *
 */
public class BroadcastVariables {

	private final static Logger log = Logger.getLogger(BroadcastVariables.class);

	public static void main(String[] args) {
		SparkConf conf = SparkUtils.getLocalConfig(BroadcastVariables.class.getSimpleName());
		log.info(String.format("access to the web interface at localhost:%s", SparkUtils.SPARK_UI_PORT));
		JavaSparkContext sparkContext = new JavaSparkContext(conf);

		List<Integer> integers = Stream.iterate(0, n -> n + 1).limit(100).collect(Collectors.toList());

		//creation
		Broadcast<List<Integer>> broadcastedIntList = sparkContext.broadcast(integers);

		//access
		broadcastedIntList.getValue();

		sparkContext.close();
	}

}

File: src/main/scala/com/joseestudillo/spark_scala/rdd/BroadcastVariables.scala
package com.joseestudillo.spark_scala.rdd

import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import org.apache.spark.sql.SQLContext
import org.slf4j.LoggerFactory
import com.typesafe.scalalogging.Logger
import org.apache.spark.broadcast.Broadcast
import com.joseestudillo.spark_scala.utils.SparkUtils

object BroadcastVariables {
  val log = Logger(LoggerFactory.getLogger(getClass.getName));

  def main(args: Array[String]) {
    val host = if (args.length > 0) args(0) else SparkUtils.LOCAL_MASTER_ID
    val appName = getClass.getName

    val conf = new SparkConf().setAppName(appName).setMaster(host)
    val sc = new SparkContext(conf)

    val integers = List.empty ++ (1 to 10)
    //creation
    val broadcastedIntList: Broadcast[List[Int]] = sc.broadcast(integers);

    //access
    broadcastedIntList.value;

    sc.stop();
  }
}